plonky2x/core/src/frontend/merkle/tendermint.rs

```diff
@@ -125,53 +163,48 @@ impl<L: PlonkParameters<D>, const D: usize> CircuitBuilder<L, D> {
             .collect_vec()
     }
 
-    pub fn get_root_from_hashed_leaves<const NB_LEAVES: usize>(
+    fn get_root_from_hashed_leaves<const MAX_NB_LEAVES: usize>(
         &mut self,
-        leaf_hashes: Vec<Bytes32Variable>,
-        leaves_enabled: Vec<BoolVariable>,
+        leaf_hashes: ArrayVariable<Bytes32Variable, MAX_NB_LEAVES>,
+        nb_enabled_leaves: Variable,
     ) -> Bytes32Variable {
-        assert!(leaf_hashes.len() == NB_LEAVES);
-        assert!(leaves_enabled.len() == NB_LEAVES);
-
         let empty_bytes = Bytes32Variable::constant(self, H256::from_slice(&[0u8; 32]));
 
         // Extend leaf_hashes and leaves_enabled to be a power of 2.
-        let padded_nb_leaves = pow(2, log2_ceil_usize(NB_LEAVES));
-        assert!(padded_nb_leaves >= NB_LEAVES && padded_nb_leaves.is_power_of_two());
+        let padded_nb_leaves = pow(2, log2_ceil_usize(MAX_NB_LEAVES));
+        assert!(padded_nb_leaves >= MAX_NB_LEAVES && padded_nb_leaves.is_power_of_two());
 
         // Hash each of the validators to get their corresponding leaf hash.
         // Pad the leaves to be a power of 2.
-        let mut current_nodes = leaf_hashes.clone();
+        let mut current_nodes = leaf_hashes.data;
         current_nodes.resize(padded_nb_leaves, empty_bytes);
 
         // Whether to treat the validator as empty.
         // Pad the enabled array to be a power of 2.
-        let mut current_node_enabled = leaves_enabled.clone();
-        current_node_enabled.resize(padded_nb_leaves, self._false());
-
-        let mut merkle_layer_size = padded_nb_leaves;
+        let mut current_node_enabled = Vec::new();
+        for i in 0..padded_nb_leaves {
+            let idx = self.constant::<Variable>(L::Field::from_canonical_usize(i));
+            current_node_enabled.push(self.lt(idx, nb_enabled_leaves));
```

Comment:
this could be expensive, you should check if it pushes you over the power of 2

Reply:
From some quick benchmarks, this doesn't seem like a significant issue. Going to leave this for now.

Reply:
We can add this optimization in the future if necessary.

Reply:
Actually nvm, going to add now.